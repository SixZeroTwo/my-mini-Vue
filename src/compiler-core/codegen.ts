import { NodeType } from "./ast"
import { CREATE_ELEMENT_BLOCK, TO_DISPLAY_STRING } from "./runtimeHelpers"

export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context
  const functionName = 'render'
  const args = ['_ctx', '_cache', '$props', '$setup', '$data', '$options']
  const signature = args.join(',')

  genFunctionPreamble(ast, context)


  push(`return function ${functionName} (${signature})`)
  push(`{ return `)
  genNode(ast.codegenNode, context)
  push(` }`)

  return {
    code: context.code
  }
}
function createCodegenContext() {
  const context = {
    code: '',
    push: (s: string) => {
      context.code += s
    },
    helper(str) {
      return `_${str}`
    }
  }
  return context
}

function genFunctionPreamble(ast, context) {
  const helpers = ast.helpers
  if (helpers.length == 0) return
  const { push } = context
  const VueBinging = 'Vue'
  const aliasHelper = helper => `${helper}: _${helper}`
  push(`const { ${helpers.map(helper => aliasHelper(helper)).join(',')} } = ${VueBinging}`)
  push('\n')
}

function genNode(node, context) {
  switch (node.type) {
    case NodeType.TEXT:
      genText(node, context)
      break
    case NodeType.INTERPLOTATION:
      genInterplotation(node, context)
      break
    case NodeType.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeType.ELEMENT:
      genElement(node, context)
      break
    case NodeType.COMPOUND:
      genCompound(node, context)
      break
  }
}

function genText(node, context) {
  const { push } = context
  push(`"${node.content}"`)
}

function genInterplotation(node, context) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(`)`)
}

function genExpression(node, context) {
  const { push } = context
  push(node.content)
}

function genElement(node, context) {
  const { push, helper } = context
  const { props, children } = node
  push(`${helper(CREATE_ELEMENT_BLOCK)}(`)
  push(`"${node.tag}"`)
  if (children.length) {
    genElementProps(props, context)
    genNodeList(children, context)
  }
  push(`)`)
}

function genNodeList(nodes, context) {
  const { push } = context
  if (nodes.length > 1) {
    push('[')
    for (let i = 0; i < nodes.length; i++) {
      const child = nodes[i]
      genNode(child, context)
    }
    push(']')
  } else {
    genNode(nodes[0], context)
  }
}

function genElementProps(props, context) {
  const { push } = context
  push(`,${props || null},`)
}

function genCompound(node, context) {
  const { children } = node
  const { push } = context
  for (let i = 0; i < children.length; i++) {
    genNode(children[i], context)
    if (i < children.length - 1) push('+')
  }
}

