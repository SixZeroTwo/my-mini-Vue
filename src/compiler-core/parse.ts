import { NodeType } from "./ast"
const enum TagType {
  START,
  END,
}

export function baseParse(content) {
  const context = createParseContext(content)
  return createRoot(parseChildren(context, []))
}

function createParseContext(content) {
  return {
    source: content
  }
}
function createRoot(children) {
  return {
    children
  }
}

function parseChildren(context, ancestors) {
  let nodes: any = []
  let node
  while (!isEnd(context, ancestors)) {
    if (context.source.startsWith('{{')) {
      node = parseInterplotation(context)

    }
    else if (/^<[a-z]+>/i.test(context.source)) {
      node = parseElement(context, ancestors)
    }
    else {
      node = parseText(context)
    }
    nodes.push(node)
  }
  return nodes
}

function parseInterplotation(context) {
  const openDelimiter = '{{'
  const closeDelimiter = '}}'
  let closeIndex = context.source.indexOf(closeDelimiter)
  let rawContentLength = closeIndex - openDelimiter.length
  //掐头
  advanceBy(context, openDelimiter.length)
  //取中
  let content = context.source.slice(0, rawContentLength).trim()
  //去尾
  advanceBy(context, rawContentLength + closeDelimiter.length)
  return {
    type: NodeType.INTERPLOTATION,
    content: {
      type: NodeType.SIMPLE_EXPRESSION,
      content,
    }
  }
}

function advanceBy(context, length) {
  context.source = context.source.slice(length)
}

function parseElement(context, ancestors) {
  //取出tag，并将头和尾去掉
  //去头
  const tag = parseTag(context, TagType.START)
  ancestors.push(tag)
  //得到children
  const children = parseChildren(context, ancestors)
  //去尾之前判断尾部标签是否与pop出来的标签一致
  const popTag = ancestors.pop()
  if (!startsWithEndTag(context.source, popTag)) {
    throw new Error(`缺少结束标签</${popTag}>`)
  }
  //去尾
  parseTag(context, TagType.END)
  return {
    type: NodeType.ELEMENT,
    tag,
    children,

  }
}

function parseTag(context, tagType) {
  const match: any = /^<\/?([a-z]+)>/i.exec(context.source)
  const tag = match[1]
  const content = match[0]
  advanceBy(context, content.length)
  return tag
}

function parseText(context: any): any {
  const symbols = ['{{', '</', '<']
  let endIndex = Infinity
  for (let sym of symbols) {
    let index = context.source.indexOf(sym)
    index == -1 ? index = Infinity : index
    endIndex = Math.min(index, endIndex)
  }
  endIndex == Infinity ? context.source.length : endIndex
  const content = context.source.slice(0, endIndex)
  advanceBy(context, endIndex)
  return {
    type: NodeType.TEXT,
    content,
  }
}

function isEnd(context, ancestors: any[]) {
  //当字符串空了或遇到parentTag时可以返回true
  const s = context.source
  if (s.startsWith('</')) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      let parentTag = ancestors[i]
      if (startsWithEndTag(s, parentTag)) return true
    }
    //结束标签开头的情况下没有匹配的开始标签
    const curEndTagArray: any = /^<\/([a-z]+)>/i.exec(s)
    const curEndTag = curEndTagArray[1]
    throw new Error(`缺少开始标签<${curEndTag}>`)
  }
  return s == ''
}

function startsWithEndTag(s, parentTag) {
  return s.slice(0, parentTag.length + 3) == `</${parentTag}>`
}
