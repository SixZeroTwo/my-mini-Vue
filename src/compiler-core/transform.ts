import { NodeType } from "./ast"
import { CREATE_ELEMENT_BLOCK, TO_DISPLAY_STRING } from "./runtimeHelpers"

export function transform(ast, options?) {
  const context = createTransformContext(ast, options ? options : {})
  traverseNodes(context.root, context)
  createCodegenNode(ast)
  ast.helpers = Array.from(context.helpers.keys())
}

function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms ? options.nodeTransforms : [],
    helpers: new Map,
    setHelpers(key) {
      if (!context.helpers.has(key)) {
        context.helpers.set(key, 1)
      }
    },
  }
  return context
}

function createCodegenNode(ast) {
  ast.codegenNode = ast.children[0]
}

function traverseNodes(node, context) {
  const { nodeTransforms } = context
  if (nodeTransforms.length) {
    for (let nodeTransform of nodeTransforms) {
      nodeTransform(node)
    }
  }

  setHelpersByType(node, context)

  traverseChildren(node, context)
}

function traverseChildren(node, context) {
  if (node.children) {
    for (let child of node.children) traverseNodes(child, context)
  }
}

function setHelpersByType(node, context) {
  switch (node.type) {
    case NodeType.INTERPLOTATION:
      context.setHelpers(TO_DISPLAY_STRING)
      break
    case NodeType.ELEMENT:
      context.setHelpers(CREATE_ELEMENT_BLOCK)
      break
  }
}

