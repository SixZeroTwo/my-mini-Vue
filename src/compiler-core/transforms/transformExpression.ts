import { NodeType } from "../ast";

export function transformExpression(node) {
  if (node.type === NodeType.INTERPLOTATION) {
    node.content = processExpression(node.content);
  }
}

function processExpression(node: any) {
  node.content = `_ctx.${node.content}`;
  return node;
}