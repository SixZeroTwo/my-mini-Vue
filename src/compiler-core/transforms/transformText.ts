import { NodeType } from "../ast";

export function transformText(node) {
  if (node.type == NodeType.ELEMENT) {
    const { children } = node
    for (let i = 0; i < children.length; i++) {
      let start = i
      let end
      const compoundNode: any = { type: NodeType.COMPOUND, children: [] }
      while (children[i] && isText(children[i])) {
        compoundNode.children.push(children[i])
        i++
      }
      end = i
      children.splice(start, end - start, compoundNode)
      i = start + 1
    }
  }
}

function isText(node) {
  return node.type == NodeType.TEXT || node.type == NodeType.INTERPLOTATION
}