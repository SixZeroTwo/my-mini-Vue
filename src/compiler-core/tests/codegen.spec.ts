import { generate } from "../codegen";
import { baseParse } from "../parse";
import { transform, } from "../transform";
import { transformExpression } from "../transforms/transformExpression";
import { transformText } from "../transforms/transformText";

describe('generate code', () => {
  test('text ast generate  code', () => {
    const ast = baseParse('hi')
    transform(ast)
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  });

  test('interplotation ast generate code', () => {
    const ast = baseParse('{{message}}')
    transform(ast, { nodeTransforms: [transformExpression] })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  });

  test('element ast generate code', () => {
    const ast = baseParse('<div></div>')
    transform(ast, { nodeTransforms: [] })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  });

  test('three types to generate code', () => {
    const ast = baseParse('<div>hi,{{message}}</div>')
    transform(ast, { nodeTransforms: [transformText] })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  });
});