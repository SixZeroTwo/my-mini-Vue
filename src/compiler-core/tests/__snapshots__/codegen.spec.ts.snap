// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate code element ast generate code 1`] = `
"const { createElementBlock: _createElementBlock } = Vue
return function render (_ctx,_cache,$props,$setup,$data,$options){ return _createElementBlock(\\"div\\") }"
`;

exports[`generate code interplotation ast generate code 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render (_ctx,_cache,$props,$setup,$data,$options){ return _toDisplayString(_ctx.message) }"
`;

exports[`generate code text ast generate  code 1`] = `"return function render (_ctx,_cache,$props,$setup,$data,$options){ return \\"hi\\" }"`;

exports[`generate code three types to generate code 1`] = `
"const { createElementBlock: _createElementBlock,toDisplayString: _toDisplayString } = Vue
return function render (_ctx,_cache,$props,$setup,$data,$options){ return _createElementBlock(\\"div\\",null,\\"hi,\\"+_toDisplayString(message)) }"
`;
